src/MultiTokenPermit.sol

```diff
@@ -0,0 +1,237 @@
+    }
+
+    /**
+     * @notice Execute multiple approved ERC721 transfers in a single transaction
+     * @dev Each transfer uses the dual-allowance system independently
+     * @param transfers Array of ERC721 transfer instructions
+     */
+    function transferFrom(
+        ERC721TransferDetails[] calldata transfers
+    ) external override {
+        uint256 transfersLength = transfers.length;
+        if (transfersLength == 0) {
+            revert EmptyArray();
+        }
+
+        for (uint256 i = 0; i < transfersLength; i++) {
+            transferFrom(transfers[i].from, transfers[i].to, transfers[i].token, transfers[i].tokenId);
+        }
+    }
+
+    /**
+     * @notice Execute multiple approved ERC1155 transfers in a single transaction
+     * @dev Each transfer uses the dual-allowance system independently
+     * @param transfers Array of multi-token transfer instructions
+     */
+    function transferFrom(
+        MultiTokenTransfer[] calldata transfers
+    ) external override {
+        uint256 transfersLength = transfers.length;
+        if (transfersLength == 0) {
+            revert EmptyArray();
+        }
+
+        for (uint256 i = 0; i < transfersLength; i++) {
+            transferFrom(
+                transfers[i].from, transfers[i].to, transfers[i].token, transfers[i].tokenId, transfers[i].amount
+            );
+        }
+    }
+
+    /**
+     * @notice Execute approved ERC1155 batch transfer for multiple token IDs to a single recipient
+     * @dev Processes each token ID individually through the dual-allowance system
+     * @param transfer Batch transfer details containing arrays of token IDs and amounts
+     */
+    function batchTransferFrom(
+        ERC1155BatchTransferDetails calldata transfer
+    ) external override {
+        uint256 tokenIdsLength = transfer.tokenIds.length;
+        if (tokenIdsLength == 0) {
+            revert EmptyArray();
+        }
+        if (tokenIdsLength != transfer.amounts.length) {
+            revert InvalidArrayLength();
+        }
+
+        // Execute batch by processing each token ID individually to leverage dual-allowance logic
+        for (uint256 i = 0; i < tokenIdsLength; i++) {
+            transferFrom(transfer.from, transfer.to, transfer.token, transfer.tokenIds[i], transfer.amounts[i]);
+        }
+    }
+
+    /**
+     * @notice Execute multiple token transfers of any type in a single transaction
+     * @dev Routes each transfer to the appropriate function based on token type
+     * @param transfers Array of multi-token transfer instructions
+     */
+    function batchTransferFrom(
+        TokenTypeTransfer[] calldata transfers
+    ) external override {
+        uint256 transfersLength = transfers.length;
+        if (transfersLength == 0) {
+            revert EmptyArray();
+        }
+
+        for (uint256 i = 0; i < transfersLength; i++) {
+            TokenTypeTransfer calldata typeTransfer = transfers[i];
```

Comment:
I don't see the point of having this TokenTypeTransfer here. Those `transferFrom` are quite generic. It's confusing sometimes tokenId 0 is treated as ERC20, and then there's the tokenType somewhereelse.